client/network/sync/src/block_relay_protocol.rs

```diff
@@ -0,0 +1,72 @@
+// Copyright Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Block relay protocol related definitions.
+
+use crate::service::network::NetworkServiceHandle;
+use futures::channel::oneshot;
+use libp2p::PeerId;
+use sc_network::request_responses::{ProtocolConfig, RequestFailure};
+use sc_network_common::sync::message::{BlockData, BlockRequest};
+use sp_runtime::traits::Block as BlockT;
+use std::sync::Arc;
+
+/// The serving side of the block relay protocol. It runs a single instance
+/// of the server task  that processes the incoming protocol messages.
+#[async_trait::async_trait]
+pub trait BlockServer<Block: BlockT>: Send {
+	/// Starts the protocol processing.
+	async fn run(&mut self);
```

Comment:
There is a problem with the current block request handler which is that even if `SyncingEngine` has rejected the peer, it's still able to download blocks because the request handler doesn't know which peers have been accepted and which have been rejected. As long as you have a connection open to the peer, you can download blocks from them.

It's not necessarily an issue with this PR but this bug has to be fixed meaning I wouldn't necessarily build on the current design of the block request handler because it has to be rewritten and coupled more tightly with `SyncingEngine`.

Reply:
>It's not necessarily an issue with this PR but this bug has to be fixed meaning I wouldn't necessarily build on the current design of the block request handler because it has to be rewritten and coupled more tightly with SyncingEngine.

Got it. Sounds like the rewrite will take a while. Since we are blocked on this change, can we get this in to make progress for now?

Reply:
Blocked how? As in you need this in Substrate ASAP?

Reply:
Blocked as in: our stack is expected to be used in bandwidth constrained environments. The duplicate gossip (once as the transactions, again as the block extrinsics) make some (gossip hungry) features in our roadmap infeasible. The savings from the dedupe would go a long way. Hope that answers your question

Reply:
I don't have any qualms about the implementation (apart from the ones that already exist there) but I would rather merge this only after we've released v0.9.42 which I believe happens soon.